asreview/hdf5_logging.py

```diff
@@ -0,0 +1,259 @@
+        i: int
+            The query number.
+        """
+        g = _result_group(self.f, query_i)
+        if "new_labels" not in g:
+            g.create_group("new_labels")
+
+        g = g['new_labels']
+
+        np_methods = np.array(list(map(np.string_, methods)))
+        _append_to_dataset('idx', idx, g, dtype=np.int)
+        _append_to_dataset('labels', labels, g, dtype=np.int)
+        _append_to_dataset('methods', np_methods, g, dtype='S20')
+
+#         bool_array = np.full(self.n_labels(), False, dtype=bool)
+#         bool_array[idx] = True
+#         self.f["labels"][bool_array] = labels
+#         for i, index in enumerate(idx):
+#             self.f["labels"][index] = labels[i]
+
+    def add_proba(self, pool_idx, train_idx, proba, query_i):
+        """Add inverse pool indices and their labels.
+
+        Arguments
+        ---------
+        indices: list, np.array
+            A list of indices used for unlabeled pool.
+        pred: np.array
+            Array of prediction probabilities for unlabeled pool.
+        i: int
+            The query number.
+        """
+        g = _result_group(self.f, query_i)
+        g.create_dataset("pool_idx", data=pool_idx, dtype=np.int)
+        g.create_dataset("train_idx", data=train_idx, dtype=np.int)
+        g.create_dataset("proba", data=proba, dtype=np.float)
+
+    def add_settings(self, settings):
+        self.settings = settings
+        self.f.attrs.pop('settings', None)
+        self.f.attrs['settings'] = np.string_(json.dumps(vars(settings)))
+
+    def n_queries(self):
+        return len(self.f['results'].keys())
+
+    def review_state(self):
+        "Get internal review state variables from hdf5 file."
+        labels = np.array(self.f['labels'][:], dtype=np.int)
+
+        train_idx = []
+        query_src = {}
+        for query_i in range(self.n_queries()):
+            try:
+                g = self.f[f'/results/{query_i}/new_labels']
+            except KeyError:
+                continue
+            new_idx = g['idx'][:]
+            methods = g['methods'][:]
+            train_idx.extend(new_idx.tolist())
+            for i in range(len(new_idx)):
+                method = methods[i]
+                idx = new_idx[i]
+                if method not in query_src:
+                    query_src[method] = []
+                query_src[method].append(idx)
+
+        train_idx = np.array(train_idx, dtype=np.int)
+        query_i = self.n_queries()-1
+        if 'labels' not in self.f[f'/results/{query_i-1}/new_labels']:
+            query_i -= 1
+
+        return labels, train_idx, query_src, query_i
+
+    def save(self):
+        """Save logs to file.
+
+        Arguments
+        ---------
+        fp: str
+            The file path to export the results to.
+
+        """
+        self.f['end_time'] = str(datetime.now())
+        self.f.flush()
+
+    def get(self, variable, query_i=None, idx=None):
+        "Get variables from hdf5 file."
```

Comment:
Docstring

Reply:
From what I found, I can put it in the base class, and the HDF5 class will inherit its docstring, right?